base/server/src/main/java/com/netscape/cms/password/PasswordChecker.java

```diff
@@ -54,17 +176,84 @@ public boolean isGoodPassword(String pwd) {
      * @return string as a reason if the password quality requirement is not met.
      */
     public String getReason(String pwd) {
-        if (pwd == null || pwd.length() == 0) {
-            EPasswordCheckException e = new EPasswordCheckException(
-                    CMS.getUserMessage("CMS_PASSWORD_EMPTY_PASSWORD"));
-
-            return e.toString();
-        } else if (pwd.length() < MIN_LEN) {
-            EPasswordCheckException e = new EPasswordCheckException(
-                    CMS.getUserMessage("CMS_PASSWORD_INVALID_LEN", "" + MIN_LEN));
+        try {
+            PasswordQuality pq = checkPassword(pwd);
```

Comment:
This is making the existing calls to isGoodPassword() and then getReason() call checkPassword() twice.  See comment at isGoodPassword.  If we change it so that isGoodPassword throws password quality failure reason in an exception it would be more efficient.  It would require some minor code changes to the callers, but I think it's only at a few places.

Reply:
Yes, it is not optimal. My idea was not to modify the code currently using the checker and update/improve in a second step but we can go fix all in one.

In my idea was not to return the reason and the output but  caching latest check so the reason can be retrieved without a check. If reason is not needed the boolean for the check is better. I'll update the code.

Reply:
@ladycfu I have updated the code. I still have to fully test, add update version script and eventually CI. But you can get a look if it is in the right direction